src/mpid/ch4/shm/ipc/src/ipc_fd.c

```diff
@@ -10,10 +10,61 @@
 /* This is defined as the max socket name length sun_path in sockaddr_un */
 #define SOCK_MAX_STR_LEN 108
 
+static int MPIDI_FD_mpi_init_hook(void);
+static int MPIDI_FD_mpi_finalize_hook(void);
+
+static int MPIDI_IPC_mpi_socks_init(void);
+static int MPIDI_IPC_mpi_fd_init(bool use_drmfd);
+static int MPIDI_IPC_mpi_fd_finalize(bool use_drmfd);
+static int MPIDI_IPC_mpi_fd_send(int rank, int fd, void *payload, size_t payload_len);
+static int MPIDI_IPC_mpi_fd_recv(int rank, int *fd, void *payload, size_t payload_len, int flags);
+
+static int ipc_fd_initialized = 0;
 static int *MPIDI_IPCI_global_fd_socks;
 static pid_t *MPIDI_IPCI_global_fd_pids;
 
-int MPIDI_FD_mpi_init_hook(void)
+int MPIDI_FD_comm_bootstrap(MPIR_Comm * comm)
+{
+    int mpi_errno = MPI_SUCCESS;
+
+    if (MPL_gpu_info.ipc_handle_type != MPL_GPU_IPC_HANDLE_SHAREABLE_FD) {
+        goto fn_exit;
+    }
+
+    if (MPIR_CVAR_CH4_IPC_ZE_SHAREABLE_HANDLE != MPIR_CVAR_CH4_IPC_ZE_SHAREABLE_HANDLE_drmfd) {
+        goto fn_exit;
+    }
+
+    MPIR_Comm *node_comm = MPIR_Comm_get_node_comm(comm);
+
+    if (node_comm->local_size == 1) {
+        goto fn_exit;
+    }
+
+    int already_initialized;
+    mpi_errno = MPIR_Allreduce(&ipc_fd_initialized, &already_initialized, 1, MPIR_INT_INTERNAL,
+                               MPI_MAX, node_comm, MPIR_ERR_NONE);
+    MPIR_ERR_CHECK(mpi_errno);
+
+    if (already_initialized) {
+        /* we can't do repeated fd sharing */
+        mpi_errno = MPI_ERR_OTHER;
+        MPIR_ERR_CHECK(mpi_errno);
+    }
```

Comment:
so will this return a runtime error in the session model when overlapping groups try to initialize GPU IPC on Intel devices? if so, does the user see that or do we just disable IPC internally?

Reply:
This is origination of the comment about impact to GPU IPC. What can cause the `already_initialized` condition and what does it mean for the user?

Reply:
We can't disable IPC at this point because there is a previous comm activated IPC already. I would say just return a runtime error. We need discuss with Intel on whether we can remove `drmfd` and always use `pidfd` (`MPIR_CVAR_CH4_IPC_ZE_SHAREABLE_HANDLE`) at this point. With `pidfd`, we can delete `ipc_fd.c`

Reply:
OK. I guess it is hard to see the impact since this function is still limited to `MPI_COMM_WORLD` in this PR. If I understand correctly, in the "true" sessions model, creating 2 comms with at least 1 overlapping process can result in an error if using Intel GPUs and `drmfd` mode. Yes, confirming with Intel which mode is preferred going forward would be good.